src/store.cc

```diff
@@ -1513,6 +1515,30 @@ StoreEntry::timestampsSet()
 
     timestamp = served_date;
 
+    debugs(20, 5, *this << " has " << describeTimestamps());
+    return true;
+}
+
+bool
+StoreEntry::updateOnNotModified(const StoreEntry &e304)
+{
+    assert(mem_obj);
+    assert(e304.mem_obj);
+
+    // update reply before calling timestampsSet() below
+    const auto &oldReply = mem_obj->freshestReply();
+    const auto updatedReply = oldReply.recreateOnNotModified(e304.mem_obj->baseReply());
+    if (updatedReply) // HTTP 304 brought in new information
+        mem_obj->updateReply(*updatedReply);
+    // else continue to use the previous update, if any
+
+    if (!timestampsSet() && !updatedReply)
+        return false;
```

Comment:
> If updatedReply is nil (nothing was updated_

nil updatedReply means that the HTTP _headers_ where not updated. timestampsSet() are responsible for _metadata_ which may (and often does) change even if headers stay the same because, IIRC, that metadata depends, in part, on current time. This method should return true if at least one of the components got change because it means we should swap the new information out to disk (or, eventually, into memory cache).

We could decline to update caches if only timestamps metadata has changed but it is not that easy to do right IIRC because you still want to update shared memory index with the new stamps so that all workers (and future requests) see them and do not revalidate again.